metagraph/src/graph/representation/succinct/boss.cpp

```diff
@@ -2623,6 +2624,72 @@ void BOSS::call_sequences(Call<std::string&&, std::vector<edge_index>&&> callbac
     }, num_threads, false, kmers_in_single_form, subgraph_mask, true);
 }
 
+// Reach all k-mers that merge into sink |edge| by following their diff paths.
+template <typename T>
+void traverse_diff_path_backward(const BOSS &boss,
+                                 edge_index edge,
+                                 T max_length,
+                                 sdsl::bit_vector *visited,
+                                 sdsl::bit_vector *terminal,
+                                 sdsl::bit_vector *dummy,
+                                 ProgressBar &progress_bar) {
+    constexpr bool async = true;
+
+    assert(max_length);
+    assert(!boss.get_W(edge) && fetch_bit(dummy->data(), edge, async));
+
+    assert(!fetch_bit(visited->data(), edge, async));
+    // mark as visited
+    set_bit(visited->data(), edge, async);
+    ++progress_bar;
+
+    std::vector<std::pair<edge_index, T /* next anchor dist */>> queue;
+
+    boss.call_incoming_to_target(boss.bwd(edge), boss.get_node_last_value(edge),
+        [&](edge_index pred) {
+            // set distance to |max_length| to make it an anchor
+            queue.emplace_back(pred, max_length);
+        }
+    );
+
+    while (queue.size()) {
+        T dist_to_anchor;
+        std::tie(edge, dist_to_anchor) = queue.back();
+        queue.pop_back();
+
+        // mark as visited
+        if (fetch_and_set_bit(visited->data(), edge, async))
```

Comment:
BTW, is it ever possible that this returns true?

Reply:
yes, it could be a loop